Can't restart with cosmos-sdk 0.38

<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺ 
v                            ✰  Thanks for opening an issue! ✰    
v    Before smashing the submit button please review the template.
v    Please also ensure that this is not a duplicate issue :)  
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Summary of Bug

I start a single node. Create one sendtx. Restart the node and it dies with a consensus failure.

This only happened after the 0.38 upgrade. Thank you to @litvintech for reporting this to me (from another v0.38.0 binary)

## Version

master branch, `697a286` and `3de4ee6` have both been shown with the issue.

## Steps to Reproduce

Setup and start node:

```
rm -rf ~/.wasmd
make install

wasmcli keys add validator
wasmcli keys add fred

wasmd init --chain-id=testing testing
wasmd add-genesis-account $(wasmcli keys show validator -a) 1000000000stake,1000000000validatortoken
wasmd gentx --name validator
wasmd collect-gentxs
wasmd start
```

Make a sendtx (in another terminal)

```
wasmcli tx send $(wasmcli keys show validator -a) $(wasmcli keys show fred -a) 98765stake -y
```

Then, kill the `wasmd` node, and start it up again.


#### Output

```
E[2020-01-27|12:20:13.030] CONSENSUS FAILURE!!!                         module=consensus err="stored minter should not have been nil" stack="goroutine 130 [running]:\nruntime/debug.Stack(0xc000da1460, 0x1114340, 0x1667790)\n\truntime/debug/stack.go:24 +0x9d\ngithub.com/tendermint/tendermint/consensus.(*State).receiveRoutine.func2(0xc000c7c700, 0x149a170)\n\tgithub.com/tendermint/tendermint@v0.33.0/consensus/state.go:613 +0x57\npanic(0x1114340, 0x1667790)\n\truntime/panic.go:679 +0x1b2\ngithub.com/cosmos/cosmos-sdk/x/mint/internal/keeper.Keeper.GetMinter(0xc000105880, 0x169cc00, 0xc00030f7e0, 0xc000105880, 0x169cc00, 0xc00030f820, 0x169cc40, 0xc00030f870, 0xc000e9be00, 0x4, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.38.0/x/mint/internal/keeper/keeper.go:57 +0x18f\ngithub.com/cosmos/cosmos-sdk/x/mint.BeginBlocker(0x16ad500, 0xc00003e120, 0x16c1bc0, 0xc000d962c0, 0xa, 0x0, 0x0, 0x0, 0x0, 0x0, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.38.0/x/mint/abci.go:11 +0x8e\ngithub.com/cosmos/cosmos-sdk/x/mint.AppModule.BeginBlock(...)\n\tgithub.com/cosmos/cosmos-sdk@v0.38.0/x/mint/module.go:130\ngithub.com/cosmos/cosmos-sdk/types/module.(*Manager).BeginBlock(0xc00011b650, 0x16ad500, 0xc00003e120, 0x16c1bc0, 0xc000d962c0, 0xa, 0x0, 0x0, 0x0, 0x0, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.38.0/types/module/module.go:297 +0x1ca\ngithub.com/cosmwasm/wasmd/app.(*WasmApp).BeginBlocker(...)\n\tgithub.com/cosmwasm/wasmd@/app/app.go:332\ngithub.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).BeginBlock(0xc000e0b7c0, 0xc000dff5e0, 0x20, 0x20, 0xa, 0x0, 0x0, 0x0, 0x0, 0x0, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.38.0/baseapp/abci.go:136 +0x469\ngithub.com/tendermint/tendermint/abci/client.(*localClient).BeginBlockSync(0xc000c1d680, 0xc000dff5e0, 0x20, 0x20, 0xa, 0x0, 0x0, 0x0, 0x0, 0x0, ...)\n\tgithub.com/tendermint/tendermint@v0.33.0/abci/client/local_client.go:231 +0x101\ngithub.com/tendermint/tendermint/proxy.(*appConnConsensus).BeginBlockSync(0xc000c26c50, 0xc000dff5e0, 0x20, 0x20, 0xa, 0x0, 0x0, 0x0, 0x0, 0x0, ...)\n\tgithub.com/tendermint/tendermint@v0.33.0/proxy/app_conn.go:69 +0x6b\ngithub.com/tendermint/tendermint/state.execBlockOnProxyApp(0x16adf00, 0xc000afc5e0, 0x16bb0a0, 0xc000c26c50, 0xc000c52c40, 0x16c3bc0, 0xc000b6c090, 0x6, 0xc000e9c560, 0x7)\n\tgithub.com/tendermint/tendermint@v0.33.0/state/execution.go:280 +0x3e1\ngithub.com/tendermint/tendermint/state.(*BlockExecutor).ApplyBlock(0xc000337ea0, 0xa, 0x0, 0xc000e9c550, 0x6, 0xc000e9c560, 0x7, 0x6, 0xc00030a2c0, 0x20, ...)\n\tgithub.com/tendermint/tendermint@v0.33.0/state/execution.go:131 +0x17a\ngithub.com/tendermint/tendermint/consensus.(*State).finalizeCommit(0xc000c7c700, 0x7)\n\tgithub.com/tendermint/tendermint@v0.33.0/consensus/state.go:1431 +0x8f5\ngithub.com/tendermint/tendermint/consensus.(*State).tryFinalizeCommit(0xc000c7c700, 0x7)\n\tgithub.com/tendermint/tendermint@v0.33.0/consensus/state.go:1350 +0x383\ngithub.com/tendermint/tendermint/consensus.(*State).enterCommit.func1(0xc000c7c700, 0x0, 0x7)\n\tgithub.com/tendermint/tendermint@v0.33.0/consensus/state.go:1285 +0x90\ngithub.com/tendermint/tendermint/consensus.(*State).enterCommit(0xc000c7c700, 0x7, 0x0)\n\tgithub.com/tendermint/tendermint@v0.33.0/consensus/state.go:1322 +0x61a\ngithub.com/tendermint/tendermint/consensus.(*State).addVote(0xc000c7c700, 0xc0001368c0, 0x0, 0x0, 0xc000d6fac0, 0x434c31, 0x149b450)\n\tgithub.com/tendermint/tendermint@v0.33.0/consensus/state.go:1819 +0xa39\ngithub.com/tendermint/tendermint/consensus.(*State).tryAddVote(0xc000c7c700, 0xc0001368c0, 0x0, 0x0, 0x8, 0xc000d6fad0, 0x4bc07d)\n\tgithub.com/tendermint/tendermint@v0.33.0/consensus/state.go:1642 +0x59\ngithub.com/tendermint/tendermint/consensus.(*State).handleMsg(0xc000c7c700, 0x168b280, 0xc000b6c1e0, 0x0, 0x0)\n\tgithub.com/tendermint/tendermint@v0.33.0/consensus/state.go:709 +0x252\ngithub.com/tendermint/tendermint/consensus.(*State).receiveRoutine(0xc000c7c700, 0x0)\n\tgithub.com/tendermint/tendermint@v0.33.0/consensus/state.go:660 +0x579\ncreated by github.com/tendermint/tendermint/consensus.(*State).OnStart\n\tgithub.com/tendermint/tendermint@v0.33.0/consensus/state.go:335 +0x13a\n"
^CE[2020-01-27|12:20:14.898] Not stopping BlockPool -- have not been started yet module=blockchain impl=BlockPool
E[2020-01-27|12:20:14.898] Stopped accept routine, as transport is closed module=p2p numPeers=0
```

Expanding the stacktrace (printing the string, not just \n):

```
goroutine 130 [running]:
runtime/debug.Stack(0xc000da1460, 0x1114340, 0x1667790)
	runtime/debug/stack.go:24 +0x9d
github.com/tendermint/tendermint/consensus.(*State).receiveRoutine.func2(0xc000c7c700, 0x149a170)
	github.com/tendermint/tendermint@v0.33.0/consensus/state.go:613 +0x57
panic(0x1114340, 0x1667790)
	runtime/panic.go:679 +0x1b2
github.com/cosmos/cosmos-sdk/x/mint/internal/keeper.Keeper.GetMinter(0xc000105880, 0x169cc00, 0xc00030f7e0, 0xc000105880, 0x169cc00, 0xc00030f820, 0x169cc40, 0xc00030f870, 0xc000e9be00, 0x4, ...)
	github.com/cosmos/cosmos-sdk@v0.38.0/x/mint/internal/keeper/keeper.go:57 +0x18f
github.com/cosmos/cosmos-sdk/x/mint.BeginBlocker(0x16ad500, 0xc00003e120, 0x16c1bc0, 0xc000d962c0, 0xa, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	github.com/cosmos/cosmos-sdk@v0.38.0/x/mint/abci.go:11 +0x8e
github.com/cosmos/cosmos-sdk/x/mint.AppModule.BeginBlock(...)
	github.com/cosmos/cosmos-sdk@v0.38.0/x/mint/module.go:130
github.com/cosmos/cosmos-sdk/types/module.(*Manager).BeginBlock(0xc00011b650, 0x16ad500, 0xc00003e120, 0x16c1bc0, 0xc000d962c0, 0xa, 0x0, 0x0, 0x0, 0x0, ...)
	github.com/cosmos/cosmos-sdk@v0.38.0/types/module/module.go:297 +0x1ca
github.com/cosmwasm/wasmd/app.(*WasmApp).BeginBlocker(...)
	github.com/cosmwasm/wasmd@/app/app.go:332
github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).BeginBlock(0xc000e0b7c0, 0xc000dff5e0, 0x20, 0x20, 0xa, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	github.com/cosmos/cosmos-sdk@v0.38.0/baseapp/abci.go:136 +0x469
github.com/tendermint/tendermint/abci/client.(*localClient).BeginBlockSync(0xc000c1d680, 0xc000dff5e0, 0x20, 0x20, 0xa, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	github.com/tendermint/tendermint@v0.33.0/abci/client/local_client.go:231 +0x101
github.com/tendermint/tendermint/proxy.(*appConnConsensus).BeginBlockSync(0xc000c26c50, 0xc000dff5e0, 0x20, 0x20, 0xa, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	github.com/tendermint/tendermint@v0.33.0/proxy/app_conn.go:69 +0x6b
github.com/tendermint/tendermint/state.execBlockOnProxyApp(0x16adf00, 0xc000afc5e0, 0x16bb0a0, 0xc000c26c50, 0xc000c52c40, 0x16c3bc0, 0xc000b6c090, 0x6, 0xc000e9c560, 0x7)
	github.com/tendermint/tendermint@v0.33.0/state/execution.go:280 +0x3e1
github.com/tendermint/tendermint/state.(*BlockExecutor).ApplyBlock(0xc000337ea0, 0xa, 0x0, 0xc000e9c550, 0x6, 0xc000e9c560, 0x7, 0x6, 0xc00030a2c0, 0x20, ...)
	github.com/tendermint/tendermint@v0.33.0/state/execution.go:131 +0x17a
github.com/tendermint/tendermint/consensus.(*State).finalizeCommit(0xc000c7c700, 0x7)
	github.com/tendermint/tendermint@v0.33.0/consensus/state.go:1431 +0x8f5
github.com/tendermint/tendermint/consensus.(*State).tryFinalizeCommit(0xc000c7c700, 0x7)
	github.com/tendermint/tendermint@v0.33.0/consensus/state.go:1350 +0x383
github.com/tendermint/tendermint/consensus.(*State).enterCommit.func1(0xc000c7c700, 0x0, 0x7)
	github.com/tendermint/tendermint@v0.33.0/consensus/state.go:1285 +0x90
github.com/tendermint/tendermint/consensus.(*State).enterCommit(0xc000c7c700, 0x7, 0x0)
	github.com/tendermint/tendermint@v0.33.0/consensus/state.go:1322 +0x61a
github.com/tendermint/tendermint/consensus.(*State).addVote(0xc000c7c700, 0xc0001368c0, 0x0, 0x0, 0xc000d6fac0, 0x434c31, 0x149b450)
	github.com/tendermint/tendermint@v0.33.0/consensus/state.go:1819 +0xa39
github.com/tendermint/tendermint/consensus.(*State).tryAddVote(0xc000c7c700, 0xc0001368c0, 0x0, 0x0, 0x8, 0xc000d6fad0, 0x4bc07d)
	github.com/tendermint/tendermint@v0.33.0/consensus/state.go:1642 +0x59
github.com/tendermint/tendermint/consensus.(*State).handleMsg(0xc000c7c700, 0x168b280, 0xc000b6c1e0, 0x0, 0x0)
	github.com/tendermint/tendermint@v0.33.0/consensus/state.go:709 +0x252
github.com/tendermint/tendermint/consensus.(*State).receiveRoutine(0xc000c7c700, 0x0)
	github.com/tendermint/tendermint@v0.33.0/consensus/state.go:660 +0x579
created by github.com/tendermint/tendermint/consensus.(*State).OnStart
	github.com/tendermint/tendermint@v0.33.0/consensus/state.go:335 +0x13a
```
____

#### For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

Adding some debug into into `Get/SetMinter` in the cosmos-sdk and then re-running gives this:

```
wasmd start
I[2020-01-27|12:54:42.638] starting ABCI with Tendermint                module=main 
SetMinter: types.Minter{Inflation:0.130000000000000000, AnnualProvisions:0.000000000000000000}
  StoreKey: KVStoreKey{0xc000bc54d0, mint}
GetMinter: types.Minter{Inflation:0.130000000000000000, AnnualProvisions:0.000000000000000000}
  StoreKey: KVStoreKey{0xc000bc54d0, mint}
SetMinter: types.Minter{Inflation:0.130000017523039605, AnnualProvisions:130000017.523039605000000000}
  StoreKey: KVStoreKey{0xc000bc54d0, mint}
I[2020-01-27|12:54:47.710] Executed block                               module=state height=1 validTxs=0 invalidTxs=0
I[2020-01-27|12:54:47.712] Committed state                              module=state height=1 txs=0 appHash=E7034D34D9122D5B6F5386275EC877BA605DCB735324300FED3D36CBED334954
GetMinter: types.Minter{Inflation:0.130000017523039605, AnnualProvisions:130000017.523039605000000000}
  StoreKey: KVStoreKey{0xc000bc54d0, mint}
SetMinter: types.Minter{Inflation:0.130000035046079271, AnnualProvisions:130000037.646079971921585420}
  StoreKey: KVStoreKey{0xc000bc54d0, mint}
I[2020-01-27|12:54:52.739] Executed block                               module=state height=2 validTxs=0 invalidTxs=0
I[2020-01-27|12:54:52.743] Committed state                              module=state height=2 txs=0 appHash=FBB9F9198A465EBD2E9DCE3E3C6D97F940F863A2F7FEE8B6A7CCA916DE47254E
GetMinter: types.Minter{Inflation:0.130000035046079271, AnnualProvisions:130000037.646079971921585420}
  StoreKey: KVStoreKey{0xc000bc54d0, mint}
SetMinter: types.Minter{Inflation:0.130000052569118999, AnnualProvisions:130000057.769121101764759960}
  StoreKey: KVStoreKey{0xc000bc54d0, mint}
I[2020-01-27|12:54:57.770] Executed block                               module=state height=3 validTxs=0 invalidTxs=0
```

Looking fine, but then in restart:

```
I[2020-01-27|12:57:21.200] starting ABCI with Tendermint                module=main 
  StoreKey: KVStoreKey{0xc000b73460, mint}
panic: stored minter should not have been nil
...
```

--- a/app/app.go
+++ b/app/app.go
@@ -146,6 +146,9 @@ func NewWasmApp(
 
 	cdc := MakeCodec()
 
+	debug := viper.GetBool(cli.TraceFlag)
+	baseAppOptions = append(baseAppOptions, bam.SetDebug(debug))
+
 	bApp := bam.NewBaseApp(appName, logger, db, auth.DefaultTxDecoder(cdc), baseAppOptions...)
 	bApp.SetCommitMultiStoreTracer(traceStore)
 	bApp.SetAppVersion(version.Version)

--- a/go.mod
+++ b/go.mod
@@ -29,3 +29,5 @@ require (
 )
 
 replace github.com/keybase/go-keychain => github.com/99designs/go-keychain v0.0.0-20191008050251-8e49817e8af4
+
+replace github.com/cosmos/cosmos-sdk => github.com/confio/cosmos-sdk v0.38.2

